File: src/csharp/Foundgine.Runtime/Hosting/Capabilities/HighAssurance.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;
using Foundgine.Core.Semantic.Security.Warrants;
using Foundgine.Runtime;

namespace Foundgine.Runtime.Capabilities;

/// <summary>
/// Turns on Foundgine's warrant-backed "high assurance" execution path: requests carrying a
/// <see cref="SecurityWarrant"/> are verified against a trusted issuer, protected from replay, and
/// revalidated again immediately before the provider runs (see <see cref="FoundgineOptions.ExecutionAuthorizationRevalidator"/>,
/// consumed by <see cref="Foundgine.Runtime.FoundgineEngine"/>).
///
/// This capability only fills in whichever of <see cref="FoundgineOptions.WarrantKeyResolver"/>,
/// <see cref="FoundgineOptions.ExpectedWarrantIssuer"/> and <see cref="FoundgineOptions.WarrantReplayStore"/>
/// the host has not already configured, using an ephemeral in-process RSA key and a local
/// file-backed replay store as the default. <b>That default is for local development and samples
/// only.</b> A real deployment must set these three explicitly - typically a KMS/HSM-backed
/// <see cref="ISecurityWarrantKeyResolver"/> and a shared (not per-instance) replay store - before
/// calling <c>Enable&lt;HighAssurance&gt;()</c>; whichever of the three are already set are left
/// untouched.
/// </summary>
public sealed class HighAssurance : IFoundgineCapability
{
    /// <summary>Issuer name used by the development default when the host has not set one.</summary>
    public const string DevIssuer = "foundgine-dev";

    public static void Configure(FoundgineCapabilityContext context)
    {
        var options = context.Options;

        options.WarrantKeyResolver ??= new EphemeralDevWarrantKeyResolver();
        options.ExpectedWarrantIssuer ??= DevIssuer;
        options.WarrantReplayStore ??= new FileSecurityWarrantReplayStore(
            Path.Combine(Path.GetTempPath(), "foundgine-dev-warrant-replay.log"));
    }

    /// <summary>
    /// Generates and caches one RSA key pair per key id for the lifetime of the process.
    /// DEV ONLY: keys never leave the process and are lost on restart, which is exactly wrong for a
    /// real deployment where warrants must stay verifiable across restarts and across instances.
    /// </summary>
    private sealed class EphemeralDevWarrantKeyResolver : ISecurityWarrantKeyResolver
    {
        private readonly ConcurrentDictionary<string, RSA> _keys = new(StringComparer.Ordinal);

        public RSA Resolve(string keyId) => _keys.GetOrAdd(keyId, static _ => RSA.Create(2048));
    }
}

/// <summary>Fluent <c>Use</c>/<c>Disable</c> surface for <see cref="HighAssurance"/>.</summary>
public static class HighAssuranceFoundgineOptionsExtensions
{
    /// <summary>Enables <see cref="HighAssurance"/>. Equivalent to <c>options.Enable&lt;HighAssurance&gt;()</c>.</summary>
    public static FoundgineOptions UseHighAssurance(this FoundgineOptions options) =>
        options.Enable<HighAssurance>();

    public static FoundgineOptions DisableHighAssurance(this FoundgineOptions options) =>
        options.Disable<HighAssurance>();
}

File: src/java/foundgine-runtime/src/main/java/com/foundgine/runtime/Capabilities/HighAssurance.java
package com.foundgine.runtime.capabilities;

import com.foundgine.core.semantic.security.warrants.*;
import com.foundgine.runtime.*;
import java.nio.file.Path;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Development capability that supplies warrant verification defaults.
 *
 * <p>
 * The generated key material is process-local and therefore intentionally
 * unsuitable for production. Production applications should configure a
 * durable/trusted key resolver and replay store before enabling this
 * capability.
 */
public final class HighAssurance implements IFoundgineCapability {
	public static final String DEV_ISSUER = "foundgine-dev";

	@Override
	public void configure(FoundgineCapabilityContext context) {
		FoundgineOptions options = context.options();
		if (options.warrantKeyResolver() == null)
			options.warrantKeyResolver(new EphemeralDevWarrantKeyResolver());
		if (options.expectedWarrantIssuer() == null)
			options.expectedWarrantIssuer(DEV_ISSUER);
		if (options.warrantReplayStore() == null)
			options.warrantReplayStore(new FileSecurityWarrantReplayStore(
					Path.of(System.getProperty("java.io.tmpdir"), "foundgine-dev-warrant-replay.log").toString()));
	}

	/** Process-local development key resolver. */
	static final class EphemeralDevWarrantKeyResolver implements ISecurityWarrantKeyResolver {
		private final ConcurrentHashMap<String, KeyPair> keys = new ConcurrentHashMap<>();

		@Override
		public java.security.interfaces.RSAPublicKey resolve(String keyId) {
			KeyPair pair = keys.computeIfAbsent(keyId, ignored -> generate());
			return (java.security.interfaces.RSAPublicKey) pair.getPublic();
		}

		private static KeyPair generate() {
			try {
				KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
				generator.initialize(2048);
				return generator.generateKeyPair();
			} catch (Exception e) {
				throw new IllegalStateException("Unable to create development RSA key.", e);
			}
		}
	}
}
